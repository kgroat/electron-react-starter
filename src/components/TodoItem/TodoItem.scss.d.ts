export const todoItem: string;
export const title: string;
export const checkboxIcon: string;
export const deleteIcon: string;
