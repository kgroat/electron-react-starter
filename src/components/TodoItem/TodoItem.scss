
@import '../../globalStyles/mixins';

.todoItem {
  @include itemBox();
  display: flex;

  .title {
    flex-grow: 1;
  }

  .checkboxIcon {
    @include icon(inherit, $hoverColor);
  }

  .deleteIcon {
    @include icon($errorText, $errorColor);
  }
}
