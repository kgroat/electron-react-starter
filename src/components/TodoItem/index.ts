
export { default, Props } from './TodoItem'
