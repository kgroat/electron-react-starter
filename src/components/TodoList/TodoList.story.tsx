
import * as React from 'react'

import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import { Todo } from 'state'

import TodoList from './'

const deleteAction = action('deleted')

const checkAction = action('checked')

const items: Todo[] = [
  {
    id: '0',
    name: 'Get milk',
    checked: false
  },
  {
    id: '1',
    name: 'Get cookies',
    checked: false
  },
  {
    id: '2',
    name: 'Do laundry',
    checked: true
  }
]

class Interactible extends React.Component<{}, { items: Todo[] }> {
  state = {
    items: [...items]
  }

  toggleChecked = (id: string, checked: boolean) => {
    this.setState(({ items }) => ({
      items: items.map(item => {
        return item.id !== id
          ? item
          : {
            ...item,
            checked
          }
      })
    }))
    checkAction(id, checked)
  }

  deleteItem = (id: string) => {
    deleteAction(id)
    this.setState(({ items }) => ({
      items: items.filter(item => item.id !== id)
    }))
  }

  render () {
    return (
      <TodoList
        items={this.state.items}
        toggleTodo={this.toggleChecked}
        deleteTodo={this.deleteItem} />
    )
  }
}

storiesOf('TodoList', module)
  .add('display', () => <TodoList items={items} toggleTodo={checkAction} deleteTodo={deleteAction} />)
  .add('interactible', () => <Interactible />)
