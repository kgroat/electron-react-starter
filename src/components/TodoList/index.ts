
export { default, Props } from './TodoList'
