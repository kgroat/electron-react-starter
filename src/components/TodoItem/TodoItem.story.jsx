
import * as React from 'react'

import { storiesOf } from '@storybook/react'

import TodoItem from './'

storiesOf('TodoItem', module)
  .add('basic example', () => <TodoItem />)
