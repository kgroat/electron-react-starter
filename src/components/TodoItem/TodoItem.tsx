
import * as React from 'react'

import { todoItem, title, checkboxIcon as checkboxIconClass, deleteIcon } from './TodoItem.scss'

export interface Props {
  name: string
  checked?: boolean
  onChange?: (checked: boolean) => void
  onDelete?: () => void
}

export default ({ name, checked, onChange, onDelete }: Props) => {
  const checkboxIcon = checked ? 'checkbox-marked-outline' : 'checkbox-blank-outline'

  return (
    <div className={todoItem}>
      <span className={title}>{name}</span>
      { onDelete ? <i onClick={() => onDelete()} className={`mdi mdi-delete ${deleteIcon}`} /> : null }
      <i onClick={() => onChange(!checked)} className={`mdi mdi-${checkboxIcon} ${checkboxIconClass}`} />
    </div>
  )
}
