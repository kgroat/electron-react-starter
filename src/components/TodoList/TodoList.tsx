
import * as React from 'react'

import { Todo } from 'state'
import TodoItem from 'components/TodoItem'

import { todoList } from './TodoList.scss'

export interface Props {
  items: Todo[]
  toggleTodo: (id: string, checked: boolean) => void
  deleteTodo: (id: string) => void
}

export default function TodoList ({ items, toggleTodo, deleteTodo }: Props) {
  return (
    <div className={todoList}>
      {
        items.map(item => (
          <TodoItem
            key={item.id}
            name={item.name}
            checked={item.checked}
            onChange={checked => toggleTodo(item.id, checked)}
            onDelete={() => deleteTodo(item.id)} />
        ))
      }
    </div>
  )
}
