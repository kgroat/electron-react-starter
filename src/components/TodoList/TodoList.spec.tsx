
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import { Todo } from 'state'
import TodoItem from 'components/TodoItem'
import TodoList from './'

function createTodo (id: string, checked: boolean): Todo {
  return {
    id,
    name: 'Get milk',
    checked
  }
}

function createTodos (count: number, checked): Todo[] {
  return Array.from(Array(count))
    .map((v, idx) => createTodo(`${idx}`, checked))
}

describe('<TodoList />', () => {
  describe('snapshots', () => {
    [0, 1, 2, 3, 4].forEach(i => {
      it(`should render correctly with ${i} unchecked todos`, () => {
        const todos = createTodos(i, false)
        const tree = create(
          <TodoList items={todos} toggleTodo={jest.fn()} deleteTodo={jest.fn()} />
        )
        expect(tree).toMatchSnapshot()
      })

      it(`should render correctly with ${i} checked todos`, () => {
        const todos = createTodos(i, true)
        const tree = create(
          <TodoList items={todos} toggleTodo={jest.fn()} deleteTodo={jest.fn()} />
        )
        expect(tree).toMatchSnapshot()
      })
    })
  })

  describe('user clicking on delete icon', () => {
    ['id1', 'id2', 'id3'].map(id => {
      it(`should call deleteTodo with ${id} once`, () => {
        const todos = [createTodo(id, false)]
        const deleteTodo = jest.fn()
        const element = shallow(
          <TodoList items={todos} deleteTodo={deleteTodo} toggleTodo={jest.fn()} />
        )
        element.find(TodoItem).props().onDelete()
        expect(deleteTodo).toHaveBeenCalledWith(id)
        expect(deleteTodo).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('user clicking on checkbox icon', () => {
    ['id1', 'id2', 'id3'].map(id => {
      it(`should call deleteTodo with ${id} and true once`, () => {
        const todos = [createTodo(id, false)]
        const toggleTodo = jest.fn()
        const element = shallow(
          <TodoList items={todos} toggleTodo={toggleTodo} deleteTodo={jest.fn()} />
        )
        element.find(TodoItem).props().onChange(true)
        expect(toggleTodo).toHaveBeenCalledWith(id, true)
        expect(toggleTodo).toHaveBeenCalledTimes(1)
      })
      it(`should call deleteTodo with ${id} and false once`, () => {
        const todos = [createTodo(id, true)]
        const toggleTodo = jest.fn()
        const element = shallow(
          <TodoList items={todos} toggleTodo={toggleTodo} deleteTodo={jest.fn()} />
        )
        element.find(TodoItem).props().onChange(false)
        expect(toggleTodo).toHaveBeenCalledWith(id, false)
        expect(toggleTodo).toHaveBeenCalledTimes(1)
      })
    })
  })
})
