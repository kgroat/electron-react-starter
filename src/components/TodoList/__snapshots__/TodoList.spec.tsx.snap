// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> snapshots should render correctly with 0 checked todos 1`] = `
<div
  className="todoList"
/>
`;

exports[`<TodoList /> snapshots should render correctly with 0 unchecked todos 1`] = `
<div
  className="todoList"
/>
`;

exports[`<TodoList /> snapshots should render correctly with 1 checked todos 1`] = `
<div
  className="todoList"
>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`<TodoList /> snapshots should render correctly with 1 unchecked todos 1`] = `
<div
  className="todoList"
>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`<TodoList /> snapshots should render correctly with 2 checked todos 1`] = `
<div
  className="todoList"
>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`<TodoList /> snapshots should render correctly with 2 unchecked todos 1`] = `
<div
  className="todoList"
>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`<TodoList /> snapshots should render correctly with 3 checked todos 1`] = `
<div
  className="todoList"
>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`<TodoList /> snapshots should render correctly with 3 unchecked todos 1`] = `
<div
  className="todoList"
>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`<TodoList /> snapshots should render correctly with 4 checked todos 1`] = `
<div
  className="todoList"
>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-marked-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`<TodoList /> snapshots should render correctly with 4 unchecked todos 1`] = `
<div
  className="todoList"
>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
  <div
    className="todoItem"
  >
    <span
      className="title"
    >
      Get milk
    </span>
    <i
      className="mdi mdi-delete deleteIcon"
      onClick={[Function]}
    />
    <i
      className="mdi mdi-checkbox-blank-outline checkboxIcon"
      onClick={[Function]}
    />
  </div>
</div>
`;
